src/cpp/session/modules/rmarkdown/NotebookAlternateEngines.cpp

```diff
@@ -452,63 +456,157 @@ Error runUserDefinedEngine(const std::string& docId,
    // always ensure we emit a 'execution complete' event on exit
    ChunkExecCompletedScope execScope(docId, chunkId);
    
-   // prepare cache output file (use tempfile on failure)
-   FilePath targetPath = module_context::tempFile("engine-cache", "txt");
-   error = prepareCacheConsoleOutputFile(docId, chunkId, nbCtxId, &targetPath);
+   // prepare cache folder
+   FilePath cachePath = notebook::chunkOutputPath(
+            docId,
+            chunkId,
+            notebook::ContextExact);
+   error = cachePath.resetDirectory();
    if (error)
-      LOG_ERROR(error);
+      return error;
    
-   // capture console output, error
-   boost::signals::scoped_connection consoleHandler =
-         module_context::events().onConsoleOutput.connect(
-            boost::bind(chunkConsoleOutputHandler, _1, _2, targetPath));
+   // helper function for emitting console text
+   unsigned int ordinal = 1;
    
-   // write input code to cache
-   error = appendConsoleOutput(
-            kChunkConsoleInput,
-            code,
-            targetPath);
-   if (error)
-      LOG_ERROR(error);
+   auto emitText = [&](const std::string& text) -> Error
+   {
+      Error error;
+
+      FilePath targetPath = notebook::chunkOutputFile(
+               docId,
+               chunkId,
+               nbCtxId,
+               ChunkOutputText);
+      
+      error = targetPath.parent().ensureDirectory();
+      if (error)
+         return error;
+
+      error = writeConsoleOutput(
+               kChunkConsoleOutput,
+               text,
+               targetPath,
+               true);
+      if (error)
+         return error;
+
+      enqueueChunkOutput(
+               docId,
+               chunkId,
+               nbCtxId,
+               ordinal++,
```

Comment:
You shouldn't need to keep track of the output ordinal yourself. If you need it, you can get it from `lastChunkOutput` (does that do what you need?)

Reply:
I tried switching to `lastChunkOutput()` but it didn't seem to work as expected for me -- consecutive plot outputs failed to display in the notebook.